metals/src/main/scala/scala/meta/internal/metals/DefinitionProvider.scala

```diff
@@ -254,22 +265,66 @@ class DestinationProvider(
     }
   }
 
-  def fromSymbol(symbol: String): Option[DefinitionDestination] = {
-    for {
-      symbolDefinition <- index.definition(Symbol(symbol))
-      if symbolDefinition.path.exists
-      destinationDoc = bestTextDocument(symbolDefinition)
-      destinationPath = symbolDefinition.path.toFileOnDisk(workspace)
-      destinationDistance =
+  def definition(
+      symbol: String,
+      source: Option[AbsolutePath]
+  ): Option[SymbolDefinition] = {
+    val allowedBuildTargets =
+      for {
+        src <- source
+        id <- buildTargets.inverseSources(src)
+      } yield buildTargets.buildTargetTransitiveDependencies(id)
+
+    val definitions = index.definitions(Symbol(symbol)).filter(_.path.exists)
+    allowedBuildTargets match {
+      case None => definitions.headOption
```

Comment:
I see this `headOption` in a lot of places. How do we ensure that the `headOption` is actually the desired one?

Reply:
It might be non desired one but at least it returns smth and works in a similar way to the current implementation.

There were two places where I wasn't sure how to pass `allowedBuildTargets` properly into this method so I decided to fix them later.

Reply:
Sounds good. For those two places, make sure to have TODOs or issues created for them then so we don't forget about them.